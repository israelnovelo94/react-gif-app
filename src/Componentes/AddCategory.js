import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {
  const [InputValue, setInputValue] = useState(''); //Enviamos string vacio para evitar el undefined y un error
  const handleInput = (e) => {
    setInputValue(e.target.value); //Actualizamos el valor del input
  }
  
  const handleSubmit = (e) => {
    e.preventDefault(); //prevenimos el refresh en la pagina
    if (InputValue.trim().length > 2) {
      setCategories(cats => [ InputValue, ...cats, ]);//Utilizamos la propiedad setCategories y añadimos el valor del input a la lista
      setInputValue('');
      
    }
  }
  return (
    <form onSubmit={handleSubmit}>
        <input 
          type="text"
          value={InputValue}
          onChange={handleInput}
        />
    </form>
  )
}
//Hacemos que la propiedad setCategorie sea obligatoria
AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}
