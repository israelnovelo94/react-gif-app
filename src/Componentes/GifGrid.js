import React from 'react'
import { useFetchGif } from '../hooks/useFetchGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {

    // Llamamos al custom hook y hacemos la desestructuracion de los datos
    const {data,loading} = useFetchGif(category);

  return (
    <>
        <h3>{category}</h3>
        {/* Hacemos una validacion para desplegar el loading en lo que carga la información */}
        {loading && <h4>Loading...</h4>}
        <div className='card-grid'>
            
            { 
                //mapeamos o recorremos la data para sacar la informacion del arreglo
                data.map( img => 
                  
                    <GifGridItem 
                    key={img.id}
                    {...img}
                    />
                    ) 
            }
            

        </div> 
    </>
  )
}
