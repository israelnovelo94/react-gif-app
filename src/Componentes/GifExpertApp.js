import React, { useState } from 'react';
import { AddCategory } from './AddCategory';
import { GifGrid } from './GifGrid';

const GifExpertApp = ( {defaultCategories = []}) => {
    

    const [categories, setCategories] = useState(defaultCategories);

    // const handleAdd = () => {
    //     setCategories( cats => [...cats, 'HunterxHunter']);
    // }
    return (
        <>
        <h2>Gix Expert App</h2>
        <hr />

        <AddCategory setCategories={ setCategories } />
        <ol>
        {
            categories.map( (category) => 

                    <GifGrid 
                        key={ category }
                        category={ category }
                    />
            )
        }
        </ol>
        </>
        
        
    )
}

export default GifExpertApp;