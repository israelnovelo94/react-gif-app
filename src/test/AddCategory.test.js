import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import { AddCategory } from '../Componentes/AddCategory';

describe('Probando el <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);
    
    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);

    })
    
    test('Probando el Snapshot', () => {

        expect(wrapper).toMatchSnapshot();
    });

    test('Probando cambiar el valor de la caja de texto', () => {

        const input = wrapper.find('input');
        const value = 'Hola Mundo';

        input.simulate('change', { target: {value}});

        expect(wrapper.find('p').text().trim()).toBe(value);
    });

    test('Simulando un submit', () => {

        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect(setCategories).not.toHaveBeenCalled();
    });

    test('llamar el setCategories y limpiar el input', () => {

        const input = wrapper.find('input');
        const value = 'Hola Mundo';

        input.simulate('change', { target: {value}});

        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith( expect.any(Function));

        expect(input.prop('value')).toBe('');
    })
})