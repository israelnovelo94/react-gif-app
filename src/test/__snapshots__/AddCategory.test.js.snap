// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el <AddCategory /> Probando el Snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
